### PR TITLE
[SYCL][COMPAT] Added USM memory

### PR DESCRIPTION
**Pull request by Alcpz**
This is the 3rd implementation PR of SYCLcompat. SYCLcompat primary goals and features are documented [here](https://intel.github.io/llvm-docs/syclcompat/README.html).

The PR includes the code that would implement the wrapper to expose memory operations, using USM. The exposed functionalities include, synchronous and asynchronous `memset`, `memcpy`, and `fill` operations; `malloc` and `free` operations; and a helper classes to query allocation type for memory pointers using SYCLcompat.

### CODE REVIEW

File: sycl/test-e2e/syclcompat/common.hpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+void instantiate_all_types(Functor &&f) {
+  auto for_each_type_call =
+      [&]<template <typename...> class T, typename... Rest>(T<Rest...> *) {
+        (f.template operator()<Rest>(), ...);
+      };
+  for_each_type_call(static_cast<types *>(nullptr));
+}
+
+#define INSTANTIATE_ALL_TYPES(types, f)                                        \
+  instantiate_all_types<types>([]<typename T>() { f<T>(); });
```
Comment (joeatodd) [inline sycl/test-e2e/syclcompat/common.hpp:37]: As we discussed offline, template parameters should be renamed here to clarify what is happening.
e.g.:
`types` -> `tuple`
`Rest` -> `Ts`
`template <typename...> class T` -> `template <typename...> class Container`

File: sycl/test-e2e/syclcompat/memory/usm_allocations.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  assert(syclcompat::detail::deduce_memcpy_direction(default_queue, s_ptr,
+                                                     d_ptr) ==
+         memcpy_direction::device_to_device);
+  assert(syclcompat::detail::deduce_memcpy_direction(default_queue, s_ptr,
+                                                     s_ptr) ==
+         memcpy_direction::device_to_device);
+}
+
+int main() {
+  using value_type_list =
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/usm_allocations.cpp]: Addressed in latest commit

File: sycl/test-e2e/syclcompat/memory/usm_allocations.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+#include <numeric>
+
+#include <sycl/sycl.hpp>
+
+#include <syclcompat/memory.hpp>
+
+#include "../common.hpp"
+#include "memory_fixt.hpp"
+
+template <typename T> void malloc() {
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/usm_allocations.cpp]: Addressed in latest commit

File: sycl/test-e2e/syclcompat/memory/memory_fixt.hpp [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+              << std::endl;
+    skip = true;
+  }
+  return skip;
+}
+
+// USM Tests Helpers
+// Fixture to set up & launch testing kernel
+template <typename T> class USMTest {
+public:
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/memory_fixt.hpp]: Addressed in latest commit

File: sycl/test-e2e/syclcompat/memory/memory_fixt.hpp [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  sycl::queue q_;
+  syclcompat::dim3 const grid_;
+  syclcompat::dim3 const thread_;
+  float *d_A;
+  float *d_B;
+  float *d_C;
+  size_t size_;
+};
+
+template <typename T> bool set_up() {
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/memory_fixt.hpp]: Addressed both comments in latest commit

File: sycl/test-e2e/syclcompat/memory/memory_fixt.hpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    assert(e1.get_info<event::command_execution_status>() ==
+           event_command_status::complete);
+  }
+
+  sycl::queue q_;
+  syclcompat::dim3 const grid_;
+  syclcompat::dim3 const thread_;
+  float *d_A;
+  float *d_B;
+  float *d_C;
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/memory_fixt.hpp]: Agreed.

File: sycl/test-e2e/syclcompat/memory/memory_fixt.hpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      assert(e2_status == event_command_status::submitted);
+    }
+
+    // Once event 2 is running, check event 1 has finished
+    while (e2_status != event_command_status::running &&
+           e2_status != event_command_status::complete) {
+      e2_status = e2.get_info<event::command_execution_status>();
+    }
+    assert(e1.get_info<event::command_execution_status>() ==
+           event_command_status::complete);
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/memory_fixt.hpp]: Addressed in latest commit as you suggested. Behavior changes a little due to that.

File: sycl/test-e2e/syclcompat/memory/memory_check.hpp [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ *  Description:
+ *    Memory content helper for the Memory functionality tests
+ **************************************************************************/
+
+#pragma once
+
+#include <cassert>
+#include <cmath>
+
+inline void check(float *h_data, float *h_ref, size_t width, size_t height,
```
[the remark this answers is not in the document]
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/memory/memory_check.hpp]: Addressed in latest commit

File: sycl/test-e2e/syclcompat/common.hpp (hunk 1 above)
Comment (Alcpz) [reply sycl/test-e2e/syclcompat/common.hpp]: Renamed to improve readability

**Review by Alcpz — COMMENTED** [review]
@joeatodd I've addressed all the issues.

**Review by joeatodd — APPROVED** [review]

**Comment by Alcpz** [comment]
@intel/llvm-gatekeepers We think this PR is ready for being merged. Thanks!.

**Pull request merged**